text/0000-third-party-web-components-support.md

```diff
@@ -0,0 +1,103 @@
+`lwc:dom="manual"` directive), the preferred way would be to enable this use case through the
+rendering of native web components. Browser support for native web components was still in the early
+stages when LWC was first implemented, but it is quite good today. In addition, many frameworks now
+support native web components as a compilation target so it makes sense to add native web component
+support as a first class citizen.
+
+## Detailed design
+
+This section covers the identified work to add native web component support.
+
+### Constructor resolution
+
+By convention, LWC implicitly resolves component constructors for every custom element in its
+template. This behavior needs to be modified to allow authors the flexibility to resolve their own
+components.
+
+#### Proposed solution
+
+LWC would provide a directive (e.g., `lwc:external`) which signals that component resolution should
+be delegated to the browser. LWC will not attempt to resolve components for custom elements tagged
+with this directive. If it happens that the component was not registered, the custom element will
+simply render as an HTMLUnknownElement, which is the default browser behavior.
+
+### Passing data
+
+By convention, custom elements are assumed to be backed by LWC components and values are set through
+their properties. Given that a native web component might expect data passed as an attribute or as a
+property, we would need to support both mechanisms. There are multiple approaches we could take
+here, and the following options are how some popular frameworks do it today. We could default to
+either attributes or properties and introduce a directive that would act as a signal to the engine
+to use one or the other, we could implement a runtime heuristic to choose one over the other, or we
+could default to attributes and set properties when setting a data structure.
+
+#### Proposed solution
+
+When passing data to a native web component, we can take a heuristical approach where if the
+property exists, the property is set. Otherwise the attribute is set. While it may be advantageous
+to require an explicit signal in the form of a directive, this approach seems to strike a good
+balance between developer experience and API semantics.
+
+### Events
+
+EventTarget polyfilled methods and properties such as `add/removeEventListener` and `target` go
+through the synthetic shadow polyfill due to those methods and properties being globally patched.
+This is the current status quo for native shadow roots and should not be an issue for supporting
+third party web components.
+
+LWC's declarative event bindings only support lowercase events. Events using anything other than
+lowercase must be listened to imperatively. This would not need to change in order to support native
+web components.
+
+### Serverside rendering (SSR)
+
+LWC's SSR implementation is tailored specifically for LWC and may not be compatible with a native

```

Comment:
I'm confused about option # 2. Could you elaborate?

For # 1, if you mean "just render `<x-foo data-bar="baz"></x-foo>` as-is, with nothing inside of the element," then that makes sense to me. It will just "hydrate" when the element is upgraded on the client-side. Not really SSR, but good enough given we don't know how to SSR other frameworks' components.

Reply:
@nolanlawson I'm talking about this:

```html
<template>
    <x-foo>
        <p></p>
    </x-foo>
</template>
```

vs 

```html
<template>
    <x-foo lwc:external>
        <p></p>
    </x-foo>
</template>
```

If `x-foo` is lwc then SSR will be able to do some work to produce the content of it... and if it has a shadow, it will produce the proper shadow as well... but if x-foo is external, then what? We can add a proprietary mechanism to infer that a shadow is needed, or they can go with the standard mechanism, which is going to be: `<x-foo lwc:external shadow-root="open">` or something like that, I don't remember.

Reply:
Doesn't rehydration take care of slotting the `<p></p>` on the client?

Reply:
Sorry, I guess I'm a little dense today. I think I'm missing an implicit assumption in this comment: `in fact, it will just produce the element as HTML, without explicit knowledge about what goes inside it... making it more tricky to determine if the shadow should be there or not`. Isn't this always the case regardless of SSR? Any custom element marked as external is simply rendered without looking for its constructor. Why do we need to know if a shadow root might be involved?

Reply:
@caridy marking this as resolved, let me know if you believe if we need to add more detail here.